src/main/java/updater/FileDownloader.java

```diff
@@ -0,0 +1,161 @@
+package updater;
+
+import javafx.application.Platform;
+import org.apache.commons.io.IOUtils;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import ui.UI;
+import util.DialogMessage;
+import util.events.Event;
+import util.events.UpdateManagerProgressWindowEvent;
+
+import java.io.*;
+import java.net.URI;
+import java.net.URLConnection;
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+
+/**
+ * A class in charge of downloading file.
+ * It has a progress tracker which can be used to show progress window.
+ */
+public class FileDownloader {
+    private static final Logger logger = LogManager.getLogger(FileDownloader.class.getName());
+
+    public static final int CONNECTION_TIMEOUT = 15000;
+    public static final int READ_CONNECTION_TIMEOUT = 30000;
+
+    private long totalFileSizeInBytes;
+
+    private final ProgressTracker progressTracker;
+
+    public FileDownloader() {
+        progressTracker = new ProgressTracker();
+    }
+
+    /**
+     * Set if current download can have its progress shown
+     *
+     * @param enableProgressWindow
+     */
+    public void setEnableProgressWindow(boolean enableProgressWindow) {
+        UI.events.triggerEvent(new UpdateManagerProgressWindowEvent(enableProgressWindow));
```

Comment:
Same thing here -- if the class knows about the update manager's concrete event, we can't stub it out, and it's implicitly coupled to whatever is triggering the event.